platform/pulp/download/single.py

```diff
@@ -0,0 +1,322 @@
+        available retries.
+
+        Args:
+            error (DownloadFailed): The raised exception.
+
+        Raises:
+            DownloadFailed: The download failed and could not be repaired.
+        """
+        try:
+            repaired = False
+            while self.retries:
+                self.retries -= 1
+                repaired = self.on_error()
+                if repaired:
+                    try:
+                        self._attempt()
+                        break
+                    except DownloadFailed as rf:
+                        repaired = False
+                        error = rf
+                else:
+                    break
+            if not repaired:
+                self.on_failed()
+                raise error
+        except Exception:
+            log.exception(_('Repair failed.'))
+            raise error
+
+    def _write(self, bfr):
+        """
+        Write downloaded content.
+
+        Args:
+            bfr (bytes): A buffer to append.
+        """
+        self.writer.append(bfr)
+        for validation in self.validations:
+            validation.update(bfr)
+
+    def _attempt(self):
+        """
+        Attempt to download.
+        """
+        with self.writer:
+            self._send()
+            self.on_reply()
+        self.on_succeeded()
+
+    def __call__(self):
+        """
+        Execute the download.
+         - _send()
+
+        Raises:
+            DownloadFailed: The download failed and could not be repaired.
+        """
+        try:
+            self.prepare()
+            self._attempt()
+        except DownloadFailed as error:
+            self._repair(error)
+
+    def _send(self):
+        """
+        Send the download.
+        This is the *main* method responsible for implementing the actual
+        download by sending a protocol specific download. The reply
+        is handled by on_reply(), on_succeeded() and on_error().
+
+        Raises:
+            DownloadFailed: The download failed and could not be repaired.
+
+        Notes:
+            Must be implemented by subclass.
+        """
+        raise NotImplementedError()
+
+    def __str__(self):
+        _id = str(id(self))[-4:]
+        description = _(
+            '{t}: id={s} url={u} path={p}'
+            ' | repair: retries={r}'
+            ' | timeout: connect={tc} read={tr}'
+            ' | auth: user={U} password={P}'
+            ' | ssl: validation={v} CA={a} key={k} certificate={c}')
+        return description.format(
+            t=type(self).__name__,
+            s=_id,
+            u=self.url,
+            p=self.path,
+            r=self.retries,
+            tc=self.connect_timeout,
+            tr=self.read_timeout,
+            U=self.user,
+            P=self.password,
+            v=self.ssl_validation,
+            a=self.ssl_ca_certificate,
+            k=self.ssl_client_key,
+            c=self.ssl_ca_certificate)
+
+
+class Context:
```

Comment:
If you're looking for a thread-safe dictionary, subclassing ``UserDict`` might be cleaner. Something like this might fit your needs:

```
from collections import UserDict
import threading


class Context(UserDict):
    def __init__(self, *args, **kwargs):
        self._lock = threading.RLock()
        super(Context, self).__init__(*args, **kwargs)

    def __getitem__(self, *args, **kwargs):
        with self._lock:
            super(Context, self).__getitem__(*args, **kwargs)

    def __setitem__(self, *args, **kwargs):
        with self._lock:
            super(Context, self).__setitem__(*args, **kwargs)

    def __delitem__(self, *args, **kwargs):
        with self._lock:
            super(Context, self).__delitem__(*args, **kwargs)
```